File: 0801-0900/0874.Walking-Robot-Simulation.cpp
#include "common.h"

int dirxy[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
      for (auto &p : obstacles) {
        is_obs.insert({p[0], p[1]});
      }
      int x = 0, y = 0, dir = 0;
      int max_dis = 0;
      for (int c : commands) {
        switch (c) {
          case -2:
            dir = (dir - 1 + 4) % 4;
            break;
          case -1:
            dir = (dir + 1) % 4;
            break;
          default:
            for (int i = 0; i < c; ++i) {
              int nx = x + dirxy[dir][0];
              int ny = y + dirxy[dir][1];
              bool fail = false;
              if (is_obs.count({nx, ny}) == 0) {
                x = nx, y = ny;
                max_dis = max(max_dis, x * x + y * y);
              }
            }
        };
      }
      return max_dis; 
    }
private:
    set<pair<int, int> > is_obs;
};

int main() {
  vector<int> commands;
  INPUT_ARRAY(commands);
  vector<vector<int> > obs;
  INPUT_ARRAY2D(obs);
  cout << Solution().robotSim(commands, obs) << endl;
  return 0;
}

File: 0801-0900/0874-walking-robot-simulation.rs
struct Solution {}

impl Solution {
    pub fn robot_sim(commands: Vec<i32>, obstacles: Vec<Vec<i32>>) -> i32 {
        use std::collections::HashSet;
        let se = obstacles.iter().map(|x| (x[0], x[1])).collect::<HashSet<_>>(); 
        let (mut x, mut y) = (0, 0);
        let mut d = 0;
        let mut ans = 0;
        let mut go = |dx, dy, step| {
            for _ in 0..step {
                let nx = x + dx;
                let ny = y + dy;
                if se.contains(&(nx, ny)) {
                    break;
                }
                x = nx;
                y = ny;
            }
            let v = x * x + y * y;
            ans = ans.max(v);
        };
        //   0
        // 1   3
        //   2
        for x in commands {
            match x {
                -2 => d = (d + 1) % 4,
                -1 => d = (d + 3) % 4,
                _  => {
                    match d {
                        0 => go(0, 1, x),
                        1 => go(-1, 0, x),
                        2 => go(0, -1, x),
                        3 => go(1, 0, x),
                        _ => (),
                    }
                },
            }
        }
        ans
    }
}

fn main() {
    let commands = vec![4,-1,4,-2,4];
    let obstacles = vec![vec![2,4]];
    println!("{}", Solution::robot_sim(commands, obstacles));
}
